apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,204 @@
+  /**
+   * Map from table name to table handler.
+   * @type {Object.<string, TableHandler>}
+   * @private
+   */
+  this.tableHandlers_ = {};
+};
+
+/**
+ * Number of ms to wait before polling each table for more data after finishing
+ * processing the last batch of data for that table.
+ * @type {number}
+ */
+var RECORD_INTERVAL = 1000; // 1 second
+
+var EventType = {
+  CREATE: 'create',
+  UPDATE: 'update',
+  DELETE: 'delete'
+};
+RecordListener.EventType = EventType;
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
+ * @param {string} tableName Name of the table to listen to.
+ * @param {function(Object, EventType)} callback Callback to call with the record
+ * and the type of event which happened to it. For create and update events,
+ * the entire new record is included. For delete events, only the id is included.
+ * @returns {boolean} true if adding the listener succeeded, or false if
+ * a listener already existed for the specified table.
+ */
+RecordListener.prototype.addListener = function (tableName, callback) {
+  if (this.tableHandlers_[tableName]) {
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
+ */
+RecordListener.prototype.reset = function () {
+  for (var tableName in this.tableHandlers_) {
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @type {{string}}
+   * @private
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @type {function}
+   *  @private
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @type {IdToJsonMap}
+   * @private
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @type {number}
+   * @private
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchNewRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchNewRecords_ = function () {
+  var req = new XMLHttpRequest();
```

Comment:
It would be nice to abstract/inject the `XMLHttpRequest` stuff to let you write unit tests around all the caching/deltas/reporting logic in here.  This has a lot in common with what we're doing in `NetSimTable`, which has a [suite of tests that fake the actual REST API](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/test/netsim/NetSimTable.js#L38).  Might be worth doing something like that here.